### DISCUSSION TITLE
Proposal for SQLite Backend code sharing

### DISCUSSION THREAD

**Discussion by insipx** [diesel-rs/diesel#4232] [Ideas]
Hello All!

For the past ~month, I've been working on an alternative [WASM SQLite backend](https://github.com/xmtp/diesel-wasm-sqlite) for diesel [crate](https://crates.io/crates/diesel-wasm-sqlite)

Interacting with external Clibs from WASM code is [tricky](https://github.com/rust-lang/rust/issues/122532). Compiling to `wasm32-unknown-emscripten` and linking with SQLite is possible, but then all `wasm-bindgen` support is lost. `diesel-wasm-sqlite` implements an alternative solution, where the SQLite WASM module is loaded separately and called from the backend. This can be a temporary solution until compiling the C SQLite dependency with `wasm-c-abi` is possible and no longer 'unstable'.

(if someone can already do that, please let me know :)

Currently lots of code in the wasm backend is copy-pasted from the diesel-backend and specialized for the `WasmSqlite` backend type. It would be great to share code with the main diesel library in order to avoid duplicating all code in the backend.

I propose creating a new feature flag `sqlite-backend` (or whatever name works) that exposes the `Sqlite` backend type, but does not import the `libsqlite3-sys` library. This would allow my backend to access all common `QueryBuilder` functions, but specialize the `WasmSqliteConnection` for a SQLite WASM module in the browser.

Another potential solution could be to split the sqlite feature flags into `sqlite-backend` and `sqlite`, enabling `sqlite-backend` in the `sqlite` feature, along with the `libsqlite3-sys` dependency. This has the advantage of less `feature = ""` declarations on items.

Signatures of common code in the backend (some of these are still private):
Essentially anything QueryBuilder-related that isn't tied to the ffi:
all of `insert_with_default_sqlite.rs`
`impl<Col, Expr> InsertValues<WasmSqlite, Col::Table>
    for DefaultableColumnInsertValue<ColumnInsertValue<Col, Expr>>`
`impl QueryFragment<WasmSqlite> for InsertOrIgnore`
`impl QueryFragment<WasmSqlite> for Replace`
`impl<T: QueryFragment<Sqlite>> QueryFragment<Sqlite> for ParenthesisWrapper<T>`
`impl QueryBuilder<WasmSqlite> for SqliteQueryBuilder`
`impl<Expr> QueryFragment<WasmSqlite, SqliteReturningClause> for ReturningClause<Expr>`
`impl QueryFragment<WasmSqlite> for LimitOffsetClause<NoLimitClause, NoOffsetClause>`
`impl<L> QueryFragment<WasmSqlite> for LimitOffsetClause<LimitClause<L>, NoOffsetClause>`
`impl<O> QueryFragment<WasmSqlite> for LimitOffsetClause<NoLimitClause, OffsetClause<O>>`
`impl<L, O> QueryFragment<WasmSqlite> for LimitOffsetClause<LimitClause<L>, OffsetClause<O>>`
`impl<'a> QueryFragment<WasmSqlite> for BoxedLimitOffsetClause<'a, WasmSqlite> `
`impl<'a> IntoBoxedClause<'a, WasmSqlite> for LimitOffsetClause<NoLimitClause, NoOffsetClause> `
`impl<'a, L> IntoBoxedClause<'a, WasmSqlite> for LimitOffsetClause<LimitClause<L>, NoOffsetClause>`
`impl<'a, O> IntoBoxedClause<'a, WasmSqlite> for LimitOffsetClause<NoLimitClause, OffsetClause<O>>`
`impl<'a, L, O> IntoBoxedClause<'a, WasmSqlite>
    for LimitOffsetClause<LimitClause<L>, OffsetClause<O>>`

All `impl ToSql<$SqlType, Sqlite> for $type`

I'd be happy to submit a PR implementing this feature flag, but wanted a go-ahead/discussion from maintainers before then, especially if there's a better way.

**Comment by weiznich** [comment]
I'm generally open to share the relevant code, but that might not be as straightforward as you believe. In the best case it would be possible to share the whole backend implementation, so that you only need to provide a custom connection implementation. That would mean that everything else would "just" work, as you could write `type Backend = Sqlite` in your implementation.

The [`Sqlite` backend](https://docs.diesel.rs/master/diesel/sqlite/struct.Sqlite.html#impl-Backend-for-Sqlite) type currently defines that the raw value type is [`SqliteValue`](https://docs.diesel.rs/master/diesel/sqlite/struct.SqliteValue.html), which is essentially a thin wrapper around the `sqlite3_value*` pointer returned from the database. That means this type is currently thighly coupled to `libsqlite3-sys`. The good news is that it doesn't expose the inner pointer type in any way, so it is possible to change that without breaking existing code. The relevant code is [here](https://github.com/diesel-rs/diesel/blob/master/diesel/src/sqlite/connection/sqlite_value.rs). I'm not 100% sure what's the best way to solve this problem.

What might be possible is to somehow store all the ffi functions used internally as function pointers in the value type, so that diesel itself could register the libsqlite3 functions there and your wasm connection could register the wasm functions there. That also would require some additional work to store the value pointer itself, although we could maybe just do that as void pointer and cast it later in some shim to the right type?
I'm not sure how that would affect performance, so this might need some benchmarking before we commit to this strategy.

**Reply by insipx** [reply]
Yeah, after writing this I realized that the SQLite backend was unique in that it held these ffi-specific types around so I closed this. I'm also not sure how wise it is to make large changes if a working `wasm-c-abi` is around the corner which would mean everything would just work in wasm anyway. But, maybe that kindof stabilization is still far off. I've been meaning to experiment linking the libsqlite3-sys library with the new unstable feature, but documentation has been a little lacking so far


I like this function pointer approach, and I'd be happy to give it a stab and see if anything comes of it.

>  I'm not sure how that would affect performance, so this might need some benchmarking before we commit to this strategy.

Are there benchmarks i'd be able to compare a preliminary implementation against in diesel which already exist?

**Reply by weiznich** [reply]
Given that this should be the only location that would need changes, I'm open to that even if the wasm-c-abi fixes are stable in the next year or so. For the other backends it's already possible to write out of tree connections, so it is desirable to have that here as well. 

There are benchmarks in the [diesel-bench](https://github.com/diesel-rs/diesel/tree/master/diesel_bench) folder. They use criterion, so it should be easy to record a baseline via `--save-baseline` and always compare to that later (via `--baseline`)

**Reply by weiznich** [reply]
Just for future reference: Here is the last update I've found for `wasm-c-abi` on the compiler side, so that sounds like it might be stable "soon".

**Reply by insipx** [reply]
Did you mean to throw a link in there? i'd be interested in it 😅

**Reply by weiznich** [reply]
Sorry, I somehow managed to not copy the link in here: https://github.com/rust-lang/rust/pull/129534
